# Load required libraries
lapply_vector <- c("DESeq2", "IHW", "tidyverse", "pheatmap", "RColorBrewer", "DOSE", 
                   "scales", "ggrepel", "cowplot", "openxlsx", "dplyr", "ashr", 
                   "factoextra", "cluster", "GenomicTools.fileHandler", "vsn", "ggvenn", 
                   "clusterProfiler", "org.At.tair.db")
lapply(lapply_vector, library, character.only = TRUE)  

# Import quantification data from count matrix (featureCounts output)
countdata <- read.table("counts.txt", header = TRUE, row.names = 1)

# Create metadata dataframe
coldata <- data.frame(
  rownames = colnames(countdata), 
  time = rep(c("24h", "24h", "24h", "2h", "2h", "2h"), times = 3), 
  condition = rep(c("661_Ac", "671_TvChi", "H2O"), each = 6),
  protocol = rep(c("H2O_2h", "CSN_MN_671", "Mock_7dpi", "CSN_MN_7dpi", "TMV_H2O", "TMV_CSN"), each = 3)
)

# Create DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countdata, colData = coldata, design = ~protocol)

# Filter out low-expression genes
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep, ]

# Differential expression analysis
dds <- DESeq(dds)

# Variance Stabilizing Transformation
vsd <- vst(dds, blind = TRUE)
vsd_mat <- assay(vsd)  # Extract VST matrix

# Compute pairwise correlation values
vsd_cor <- cor(vsd_mat)

# Generate heatmap
pheatmap(vsd_cor)

# MA plots
plotMA(dds, ylim = c(-5, 5))
resBigFC <- results(dds, lfcThreshold = 1, altHypothesis = "greaterAbs")
plotMA(resBigFC, ylim = c(-5, 5))
abline(h = c(-1, 1), lwd = 5)

# Principal Component Analysis (PCA)
plotPCA(vsd, intgroup = "protocol")

# Dispersion estimates
plotDispEsts(dds)

# Extract differential expression results
Contrast <- c("protocol", "H2O_2h", "CSN_MN_671")
comparison <- "CSN_MN_671_H2O_2h"
comparison_title <- "CSN_MN_671_H2O_2h"

# Shrink log2 fold changes to stabilize variance
res_tableOE_unshrunken <- results(dds, contrast = Contrast, alpha = 0.05)
res_tableOE <- lfcShrink(dds = dds, contrast = Contrast, type = "ashr", res = res_tableOE_unshrunken)

# MA plots for shrunken and unshrunken results
plotMA(res_tableOE_unshrunken, ylim = c(-2, 2))
plotMA(res_tableOE, ylim = c(-2, 2))

# Convert results to tibble format
res_tableOE_tb <- res_tableOE %>%
  data.frame() %>%
  rownames_to_column(var = "gene") %>%
  as_tibble()

# Filter significant genes
sigOE <- res_tableOE_tb %>%
  filter(padj < 0.05 & (log2FoldChange > 2 | log2FoldChange < -2))

# Save significant gene list to file
write.table(sigOE, file = "sigOE1_H2O_2hvsCSN_MN_671_1.txt", sep = "\t", quote = FALSE, col.names = NA)
